firebase-vertexai/src/main/kotlin/com/google/firebase/vertexai/type/AudioHelper.kt

```diff
@@ -17,125 +17,160 @@
 
 import android.Manifest
+import android.media.AudioAttributes
 import android.media.AudioFormat
 import android.media.AudioManager
 import android.media.AudioRecord
 import android.media.AudioTrack
 import android.media.MediaRecorder
 import android.media.audiofx.AcousticEchoCanceler
 import androidx.annotation.RequiresPermission
+import com.google.firebase.vertexai.common.util.readAsFlow
 import kotlinx.coroutines.flow.Flow
-import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.emptyFlow
 
+/**
+ * Helper class for recording audio and playing back a separate audio track at the same time.
+ *
+ * @see AudioHelper.build
+ * @see LiveSession.startAudioConversation
+ */
 @PublicPreviewAPI
-internal class AudioHelper {
+internal class AudioHelper(
+  /** Record for recording the System microphone. */
+  private val recorder: AudioRecord,
+  /** Track for playing back what the model says. */
+  private val playbackTrack: AudioTrack,
+) {
+  private var released: Boolean = false
 
-  private lateinit var audioRecord: AudioRecord
-  private lateinit var audioTrack: AudioTrack
-  private var stopRecording: Boolean = false
+  /**
+   * Release the system resources on the recorder and playback track.
+   *
+   * Once an [AudioHelper] has been "released", it can _not_ be used again.
+   *
+   * This method can safely be called multiple times, as it won't do anything if this instance has
+   * already been released.
+   */
+  fun release() {
+    if (released) return
+    released = true
 
-  internal fun release() {
-    stopRecording = true
-    if (::audioRecord.isInitialized) {
-      audioRecord.stop()
-      audioRecord.release()
-    }
-    if (::audioTrack.isInitialized) {
-      audioTrack.stop()
-      audioTrack.release()
-    }
+    recorder.release()
+    playbackTrack.release()
   }
 
-  internal fun setupAudioTrack() {
-    audioTrack =
-      AudioTrack(
-        AudioManager.STREAM_MUSIC,
-        24000,
-        AudioFormat.CHANNEL_OUT_MONO,
-        AudioFormat.ENCODING_PCM_16BIT,
-        AudioTrack.getMinBufferSize(
-          24000,
-          AudioFormat.CHANNEL_OUT_MONO,
-          AudioFormat.ENCODING_PCM_16BIT
-        ),
-        AudioTrack.MODE_STREAM
-      )
-    audioTrack.play()
+  /**
+   * Play the provided audio data on the playback track.
+   *
+   * Does nothing if this [AudioHelper] has been [released][release].
+   */
+  fun playAudio(data: ByteArray) {
+    if (released) return
+
+    playbackTrack.write(data, 0, data.size)
   }
 
-  internal fun playAudio(data: ByteArray) {
-    if (!stopRecording) {
-      audioTrack.write(data, 0, data.size)
-    }
+  /**
+   * Pause the recording of the microphone, if it's recording.
+   *
+   * Does nothing if this [AudioHelper] has been [released][release].
+   *
+   * @see resumeRecording
+   */
+  fun pauseRecording() {
+    if (released || recorder.recordingState == AudioRecord.RECORDSTATE_STOPPED) return
+
+    recorder.stop()
```

Comment:
this can throw an exception, shouldn't we release the resources if so?

Reply:
Oh good catch, done.